src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,236 @@
+
+use anyhow::{anyhow, Context, Result};
+use kata_sys_util::rand::RandomBytes;
+use tokio::sync::Mutex;
+
+use crate::{
+    BlockConfig, Device, DeviceConfig, Hypervisor, KATA_BLK_DEV_TYPE, KATA_MMIO_BLK_DEV_TYPE,
+    VIRTIO_BLOCK_MMIO, VIRTIO_BLOCK_PCI,
+};
+
+use super::util::{get_host_path, get_virt_drive_name};
+pub type ArcBoxDevice = Arc<Mutex<Box<dyn Device>>>;
+
+/// block_index and released_block_index are used to search an available block index
+/// in Sandbox.
+///
+/// @block_index generally default is 1 for <vdb>;
+/// @released_block_index for blk devices removed and indexes will released at the same time.
+#[derive(Clone, Debug, Default)]
+struct SharedInfo {
+    block_index: u64,
+    released_block_index: Vec<u64>,
+}
+
+impl SharedInfo {
+    fn new(index: u64) -> Self {
+        SharedInfo {
+            block_index: index,
+            released_block_index: vec![],
+        }
+    }
+
+    // declare the available block index
+    fn declare_device_index(&mut self) -> Result<u64> {
+        let current_index = if let Some(index) = self.released_block_index.pop() {
+            index
+        } else {
+            self.block_index
+        };
+        self.block_index += 1;
+
+        Ok(current_index)
+    }
+
+    fn release_device_index(&mut self, index: u64) {
+        self.released_block_index.push(index);
+        self.released_block_index.sort_by(|a, b| b.cmp(a));
+    }
+}
+
+// Device manager will manage the lifecycle of sandbox device
+pub struct DeviceManager {
+    devices: HashMap<String, Arc<Mutex<Box<dyn Device>>>>,
+    hypervisor: Arc<dyn Hypervisor>,
+    shared_info: SharedInfo,
+}
+
+impl DeviceManager {
+    pub fn new(hypervisor: Arc<dyn Hypervisor>) -> Result<Self> {
+        let devices = HashMap::<String, Arc<Mutex<Box<dyn Device>>>>::new();
+        Ok(DeviceManager {
+            devices,
+            hypervisor,
+            shared_info: SharedInfo::new(1),
+        })
+    }
+
+    pub async fn new_device(&mut self, device_config: &DeviceConfig) -> Result<String> {
+        let device_id = if let Some(dev) = self.find_device(device_config).await {
+            dev
+        } else {
+            self.create_device(device_config)
+                .await
+                .context("failed to create device")?
+        };
+        Ok(device_id)
+    }
+
+    pub async fn try_add_device(&mut self, device_id: &String) -> Result<()> {
+        let device = self
+            .devices
+            .get_mut(device_id)
+            .context("failed to find device")?;
+        // increase attach count, skip attach the device if the device is already attached
+        let need_skip = device
+            .lock()
+            .await
+            .increase_attach_count()
+            .await
+            .context("failed to increase attach count")?;
+        if need_skip {
+            return Ok(());
+        }
+        let result = device.lock().await.attach(self.hypervisor.as_ref()).await;
```

Comment:
we do `device.lock()` for each time we use `device` will make the lock more fine-grained, so It will have better concurrency

Reply:
Yeah, we should take a balance here. I would like to hear other people's opinions. WDYT? @lifupan @quanweiZhou

Reply:
> Yeah, we should take a balance here. I would like to hear other people's opinions. WDYT? @lifupan @quanweiZhou

Since those operations are near one by one, I think it's better to brace them in one lock.

Reply:
Agree with @YushuoEdge and @lifupan.

Reply:
fixed